### PR TITLE
WIP: Adds support for connecting to BIG-IPs on off-numbered ports

### LINKED ISSUE #347: Add port option
I test BIG-IPs through a vagrant port forward, so the port is on an off number

ex
port 10443 forwarded to port 443 via vagrant

To support thing in the SDK I could make the hostname "localhost:10443", but supporting a port option that defaults to 443 sounds like it would make more sense.

I think an additional enhancement would be required in the icontrol rest library

### PR DESCRIPTION
**Pull request by caphrim007**
Issues:
Fixes #347

Problem:
The f5-sdk makes the assumption that all BIG-IPs are accessed on
port 443. While this may be true in most cases, there are cases
where the BIG-IP's web ui may be accessed through a port-forward
and therefore the f5-sdk would need to specify the forwarded port
instead of assuming the default.

Analysis:
I added a port argument to the BigIP constructor. By default this
value is the standard port that the web ui is accessed on; 443.

Whether the port is specified or not, the string formatting performed
for the self._meta_data['uri'] value will include the port value.
The new format of the 'uri' value will resemble the following

```
https://FakeHostName:443/mgmt/tm
```

If a port is specified, for instance 10443, it would then look like

```
https://FakeHostName:10443/mgmt/tm
```

This change allows BIG-IPs to be accessed on off-numbered ports

Tests:
- FakeBigIPWithPort
- test_meta_data_uri

### CODE REVIEW

**Comment by caphrim007** [comment]
@swormke @zancas

File: f5/bigip/__init__.py [hunk 1]
```diff
@@ -43,6 +44,7 @@ def __init__(self, hostname, username, password, **kwargs):
         self._meta_data = {
             'allowed_lazy_attributes': allowed_lazy_attrs,
             'hostname': hostname,
+            'port': port,
             'uri': 'https://%s/mgmt/tm/' % hostname,
```
Comment (swormke) [inline f5/bigip/__init__.py]: Should the URI here be `'https://%s:%s/mgmt/tm/' % (hostname, port)`

**Comment by swormke** [comment]
Can't believe we didn't have this from the beginning!  Thanks for adding.  I had one comment that I put inline about the _meta_data['uri'] in think that when we build that we need to also include the port right?

I'll let @zancas take a look too since he is closer to the code these days than I am.

File: f5/bigip/__init__.py (hunk 1 above)
Comment (caphrim007) [inline f5/bigip/__init__.py]: I originally thought that it should, but there is code further down in the library that uses the uri in some ways that I wasn't able to easily grab the port from....or something like that.

Lemme see if I can find my reasoning behind doing that. I think it was in the resource.py file.

**Comment by caphrim007** [comment]
@swormke Now that I understand better how the SDK works, let me try again by tweaking the uri field

**Comment by caphrim007** [comment]
@swormke @zancas I've made updates as @swormke suggested. It still seems to work on my end I think (if I havent hosed by virtualenv enough :-P) Have a look

File: f5/bigip/resource.py [hunk 2]
```diff
@@ -532,15 +532,15 @@ def _activate_URI(self, selfLinkuri):
         :param selfLinkuri: the server provided selfLink (contains localhost)
         :raises: URICreationCollision
         """
-        # hostname local alias
-        hostname = self._meta_data['bigip']._meta_data['hostname']
+        # netloc local alias
+        uri = urlparse.urlsplit(self._meta_data['bigip']._meta_data['uri'])
```
Comment (zancas) [inline f5/bigip/resource.py:536]: Nice!  +1    So urlsplit is returning a NamedTuple?  I did not realize...   awesome.

File: f5/bigip/test/test___init__.py [hunk 3]
```diff
@@ -42,3 +50,8 @@ def test___get__attr(FakeBigIP):
     with pytest.raises(AttributeError):
         FakeBigIP.this_is_not_a_real_attribute
     assert FakeBigIP.hostname == 'FakeHostName'
+
+
+def test_meta_data_uri(FakeBigIPWithPort):
```
Comment (zancas) [inline f5/bigip/test/test___init__.py]: This is slightly more specific than a just a test of "meta_data_uri" right, it's a test of the non-default port number, so let's rename it to something that includes the words "port", "number", and "default".

File: f5/bigip/__init__.py [hunk 4]
```diff
@@ -30,7 +30,7 @@
 
 class BigIP(OrganizingCollection):
     """An interface to a single BIG-IP"""
-    def __init__(self, hostname, username, password, **kwargs):
+    def __init__(self, hostname, username, password, port=443, **kwargs):
```
Comment (zancas) [inline f5/bigip/__init__.py]: We're using the keyword args default params pattern here, soo...

Instead of having `port=443` in the param list for the constructor, move it down into the "kwargs.pop"s.

So the whole param list should be:

`def __init__(self, hostname, username, password, **kwargs):`

and then...   see next line note...

File: f5/bigip/__init__.py [hunk 5]
```diff
@@ -30,7 +30,7 @@
...
         timeout = kwargs.pop('timeout', 30)
         allowed_lazy_attrs = kwargs.pop('allowed_lazy_attributes',
                                         allowed_lazy_attributes)
```
Comment (zancas) [inline f5/bigip/__init__.py:37]: Add this:

`port = kwargs.pop('port', 443)`

Comment (zancas) [inline f5/bigip/__init__.py:37]: Then even though it's not explicitly about the port number add:

```
if kwargs:
    raise TypeError('Unexpected kwargs: %r' % kwargs)
```

**Comment by zancas** [comment]
@caphrim007 
OK, I've added line notes with all my comments.

For the most part it looks great, and I'm excited to see that the return type of urlsplit is a NamedTuple...  cool!    Nice improvement.

I'd like us to continue to conform to a single pattern for default params.  We've decided on the `keyword-only` default params (From "Effective Python), so let's do the teeny refactor needed to move the `param=443` from a param to a kwarg key-value.   

See line notes for details --^

Once that's done I'll merge.

**Comment by caphrim007** [comment]
@zancas changes incoming
